crates/api/src/func.rs

```diff
@@ -995,3 +829,194 @@ impl<T: WasmTy> WasmRet for Result<T, Trap> {
         T::store(abi, ptr);
     }
 }
+
+/// Internal trait implemented for all arguments that can be passed to
+/// [`Func::wrap`].
+///
+/// This trait should not be implemented by external users, it's only intended
+/// as an implementation detail of this crate.
+pub trait IntoFunc<Params, Results> {
+    #[doc(hidden)]
+    fn into_func(self, store: &Store) -> Func;
+}
+
+/// A structure representing the *caller's* context when creating a function
+/// via [`Func::wrap`].
+///
+/// This structure can be taken as the first parameter of a closure passed to
+/// [`Func::wrap`], and it can be used to learn information about the caller of
+/// the function, such as the calling module's memory, exports, etc.
```

Comment:
It'd be good to mention somewhere that this is intended to be a temporary mechanism, and that we're hoping interface types will provide better ways to do this in the future.